### ISSUE TITLE
Digital Forensic Data - WAL & Journal Records

### ISSUE THREAD

**Issue by TroySchnack** [sqlitebrowser/sqlitebrowser#1855]
Describe the new feature
--------------------------
Ability to view records/data from a database's Journal or WAL file. This process is used often by digital forensic investigators to locate deleted information. The feature would offer the user the option to load and browse for the location of the Journal or WAL file and choose whether or not to include their data.

Additional info
---------------
Please answer these questions before submitting your feature request.

No issue

Attached is sample database file and journal from an Android device.

[db and journal sample.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/3103655/db.and.journal.sample.zip)

**Comment by justinclift** [comment]
With the sample database, what should we be looking for that isn't showing up at the moment? :smile:

**Comment by TroySchnack** [comment]
Sorry about that. There is a deleted sms message record.

“It is.  Good thing the cell companies only keep the messages for a short time.”

[cid:image001.png@01D4F8EF.63CAE320]




MCFE | EnCE | ACE | CMFF | CCO | CCPA


Troy Schnack
Computer System Administrator
Federal Public Defender
818 Grand Blvd, Suite 300
Kansas City, MO 64106

Main:   (816) 471-8282
Direct:   (816) 325-2444
Mobile:  (816) 200-4758


*This e-mail contains PRIVILEGED and CONFIDENTIAL information intended only for the use of the addressee(s) named above.  If you are not the intended recipient of this e-mail, or an authorized employee or agent responsible for delivering it to the intended recipient, you are hereby notified that any dissemination or copying of this e-mail is strictly prohibited.  If you have received this e-mail in error, please notify us by reply e-mail. Thank you for your cooperation.

**Comment by justinclift** [comment]
Thanks @TroySchnack. :smile:

GitHub doesn't attach images added via email.  You ok to manually drop that image in via the GitHub webUI?

**Comment by justinclift** [comment]
Looking at our code which does the opening of the database here:

https://github.com/sqlitebrowser/sqlitebrowser/blob/7b6bc013cddf3652345c1ab3f5458c2a446e1546/src/sqlitedb.cpp#L150-L155

To get it initially working, we might just need to adjust the parameters to the [`sqlite3_open_v2()`](https://sqlite.org/c3ref/open.html) call, so it notices the journal file.

Probably something to do with the flags. :smile:

We should probably also add some checking for a `<filename>-journal` file too (ditto for a WAL file) in the same directory too, prior to opening the database.  That could then do special handling in the open piece, to ensure the journal/wal records are incorporated.

**Comment by justinclift** [comment]
@MKleusberg @mgrojo Thoughts? :smile:

**Comment by TroySchnack** [comment]
Majority of DB managers wouldn’t want to view deleted content, so it should probably be an option to enable those records or not for browsing data or SQL statements.




MCFE | EnCE | ACE | CMFF | CCO | CCPA


Troy Schnack
Computer System Administrator
Federal Public Defender
818 Grand Blvd, Suite 300
Kansas City, MO 64106

Main:   (816) 471-8282
Direct:   (816) 325-2444
Mobile:  (816) 200-4758


*This e-mail contains PRIVILEGED and CONFIDENTIAL information intended only for the use of the addressee(s) named above.  If you are not the intended recipient of this e-mail, or an authorized employee or agent responsible for delivering it to the intended recipient, you are hereby notified that any dissemination or copying of this e-mail is strictly prohibited.  If you have received this e-mail in error, please notify us by reply e-mail. Thank you for your cooperation.

**Comment by justinclift** [comment]
Yeah, I'm not sure if/how we'd do it either.  I'm thinking that since we're using the standard SQLite library, and from it's point of view the data is deleted... there's not much we could do.

But, I'm not 100% sure.  It's an interesting use case to think through anyway, and there might be some other approach we can come up with. :smile:

Radical alternative idea... someone writes a drop-in-replacement SQLite library (or modifies the standard one), so it picks up deleted records too?  That way we could compile DB4S to use that, and it'd work...  hmmm... just thinking out loud. :smile:
